### ISSUE TITLE
Typescript port and refactoring 

### ISSUE THREAD

**Issue by esjewett** [crossfilter/crossfilter#84]
As discussed with @tehsenaus over on #81, Crossfilter could use some significant refactoring. Porting to a typed language would also help with ensuring things are working as we expect and cutting down on bugs, so moving to Typescript may be helpful in this endeavor. This issue is to discuss those possible changes and track work related to them.

Do people have thoughts on this proposal?

**Comment by supercargo** [comment]
I ended up stumbling upon this issue because I was looking for info on type definitions for the community version of crossfilter.  I think a typescript port would be a great way to go.  

Adding static types to the existing API seems like a reasonable first step.  `@types/crossfilter` is a start, but doesn't have any of the 1.4 API.  I was considering authoring a `crossfilter2` port for DefinitelyTyped, but getting types added natively to a version of crossfilter would be even better.

**Comment by esjewett** [comment]
@supercargo That does seem like it would be a good first step. Just as `crossfilter2` (on the 1.x major release) is a purely additive fork of `crossfilter`, I imagine that the types would be purely additive as well.

What would our options be for getting types added directly in the code base? A full Typescript rewrite, or is there a way to annotate the code base in place?

**Comment by supercargo** [comment]
@esjewett Initially you can do as little as packaging a handwritten `.d.ts` file based on [the one in `@types/crossfilter`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/crossfilter/index.d.ts).  It would probably make sense to also have the ts compiler process this file to make sure it is valid as part of tests, but I'm not sure what the current build process looks like.

As the codebase migrates to TypeScript, the type definitions can be migrated into regular `.ts` files and `.d.ts` will be generated as output of the compiler.

**Comment by gordonwoodhull** [comment]
How does Typescript handle arguments that could be one of two types? Because the keys in crossfilter must be either numbers or strings. (Or dates, but those coerce to numbers.)

The terse crossfilter docs just say they must be naturally-ordered. Here's a little wiki entry I wrote about it, somewhat relevant:

https://github.com/crossfilter/crossfilter/wiki/Crossfilter-Gotchas#natural-ordering-of-dimension-and-group-values

People also supply arrays as keys, either using the tag-like [isArray](https://github.com/crossfilter/crossfilter/wiki/API-Reference#dimension_with_arrays) functionality, or without it, which causes keys to be coerced to strings (which often works but can lead to weird bugs).

The other place I see where strict typing might not work is the data supplied to the crossfilter constructor or `.add()`, which must be an array, but an array of anything that works with the key and reduce functions. 

I guess those parts would have to be left untyped, but the rest of the interface could be specified?

**Comment by tehsenaus** [comment]
@gordonwoodhull TypeScript supports generics and union types, so all you've described would work quite well. E.g.:

```
type Key = number | string | etc...;
```

```
interface Crossfilter<T> {
    add(data: T []);
}
```

If there's a demand then I'm happy to give this refactor a go when I have some time, it should be a fun project. And I think it can be done without any breaking API changes.

**Comment by paulsjv** [comment]
I would love this feature as I am trying to add crossfilter to a project I am working on in Angular 4.  I would much rather have the @types/crossfilter as a short-term solution to get it working with Angular 4 and a longer-term solution of converting it to typescript with the @types built into the code.

I have little experience with TypeScript but would be willing to learn and help out when I can.

**Comment by esjewett** [comment]
I'd love to see this as well but have no time to work on it until January, probably. If you all want to put together either a `@types/crossfilter2` project, or a pull request (or both!), go for it!

**Comment by tehsenaus** [comment]
I started work on the TypeScript port - I reckon it's about 50% there. But in the meantime it should be very straightforward to add types to this package. I can look into this.

**Comment by paulsjv** [comment]
I just started on the TypeScript `@types/` declaration today.

**Comment by adrm** [comment]
How is this issue progressing? Right now, the absence of types makes it impossible to use crossfilter2 with `"strict": true` option in the typescript compiler. What is the recommended workaround, if any?

**Comment by paulsjv** [comment]
Here's the PR I've made from my fork.  

https://github.com/crossfilter/crossfilter/pull/86

I haven't had time to work on it since I created the PR due to work, holidays, and illness.  However, feel free to pick up where I left off!  I won't have time to pick it back up until after our yearly review cycle is over in February.

**Comment by paulsjv** [comment]
I know that they type definition was released with 1.4.5.  However, when trying to install the type definition using `npm install --save @types/crossfilter2` I get a 404 not found.  I know the type definition is there in the root directory and in the package.json.  I would just like to be able to install it without having to go the custom install route.  Is there any way to get this resolved?

**Comment by esjewett** [comment]
@paulsjv Can you give an example of your setup? Since the types are bundled with the main package and are referenced in the package.json, my impression was that the typescript compiler should pick them up. The types aren't published on DefinitelyTyped as far as I'm aware, so trying to install `@types/crossfilter2` will not work.

**Comment by paulsjv** [comment]
@esjewett Yes!  I'm actually working on an implementation of it for the DefinitelyTyped repo to share here.  This way all you'll have to do is for DefinitelyTyped and add what's in my fork to yours then create a PR on their repo!  Although, in doing this I'm learning that the d.ts file here is not exactly what is needed so I'll be making some changes to it and delivering a PR here as well.

**Comment by fnberta** [comment]
There is absolutely no need for this @paulsjv. Just install the normal crossfilter2 package and the types are there. You can have typescript types directly in the npm package. DefinitelyTyped is only needed if package maintainers dont want to have the types in the main npm package. Having them on both would just confuse everybody.

**Comment by paulsjv** [comment]
Thanks, @fnberta I got it working correctly in my app.  Being new to Typescript, I'm wondering if there should be some documentation on how to get this to work with a user's app?  I finally got it working because I had the wrong import and needed `import * as cf from 'crossfilter2'`.

None the less, I guess the question becomes does the community want to publish the type definitions to DefinitelyTyped or not?  I'd like it because it seems easy but you're right the pitfall would be to have to keep it up to date until the project moves to typescript and the d.ts can be auto-generated.  I think in the short term maybe some documentation would suffice?

**Comment by paulsjv** [comment]
Then maybe a line of documentation would be appropriate?

**Comment by phyratokar** [comment]
I think a bit of documentation or just an example of how to import and use it in typescript would be helpful.
After `import * as cf from 'crossfilter2';` whats the type of `cf(data)`?

**Comment by esjewett** [comment]
That (a documentation update) would be excellent. If anyone wants to write something for the docs, just open another issue with what you want to add and I'll be happy to merge it in.

**Comment by TijmenW** [comment]
@tehsenaus, is anyone currently working on a TypeScript port?

**Comment by hornta** [comment]
I plan on making some effort on this. I think it can be wise to gradually port this to TS because it's doable when building with Rollup.  TypeScript has this magical type called `unknown` which is a perfect substitute for well unknown data types which is perfect to use when porting. However the manual type declaration file will have to remain until every public API is typed.

I don't believe in trying to make it all in one go but what do you think?

**Comment by gordonwoodhull** [comment]
I’m sorry, I'm not interested in reviewing a change to TS.

**Comment by gordonwoodhull** [comment]
Sorry, wrong ticket. If anyone steps up as maintainer they should feel free to make this change!

**Comment by gordonwoodhull** [comment]
I want to emphasize that this is a good idea - I just don't have time to review it, and don't have much experience with TS.

One of your options is to fork crossfilter now, do your prettier and porting work in your own fork, and see if anyone steps up to maintain this repo who would be willing to review and merge your changes. Or maybe you will find that the code is not that bad, and you want to step up yourself. :)
